include/node.tcc

```diff
@@ -558,3 +559,43 @@ void mpm::Node<Tdim, Tdof,
                                       change_in_momenta, Tdim);
   }
 }
+
+//! Compute multimaterial separation vector
+template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
+void mpm::Node<Tdim, Tdof,
+               Tnphases>::compute_multimaterial_separation_vector() {
+  // iterate over all materials in the material_ids set, update the
+  // displacements and calculate the displacement of the center of mass for this
+  // node
+  std::lock_guard<std::mutex> guard(node_mutex_);
+  for (auto mitr = material_ids_.begin(); mitr != material_ids_.end(); ++mitr) {
+    const Eigen::Matrix<double, Tdim, 1> material_displacement =
+        property_handle_->property("displacements", prop_id_, *mitr, Tdim);
+    const Eigen::Matrix<double, 1, 1> material_mass =
+        property_handle_->property("masses", prop_id_, *mitr);
+
+    // displacement of the center of mass
+    displacement_ += material_displacement / mass_(0, 0);
+    // assign nodal-multimaterial displacement by dividing it by this material's
+    // mass
+    property_handle_->assign_property(
+        "displacements", prop_id_, *mitr,
+        material_displacement / material_mass(0, 0), Tdim);
+  }
+
+  // iterate over all materials in the material_ids to compute the separation
+  // vector
```

Comment:
Does this have to be computed after the displacement in previous for, could we merge these two for loops?

Reply:
They can't be merged because the displacement of the center of mass needs to be calculated in the first loop before it can be used in the second loop to calculate the separation vector.